gpt_engineer/ai.py

```diff
@@ -23,72 +39,134 @@ class TokenUsage:
         ]
-
         return self.next(messages, step_name=step_name)
 
-    def fsystem(self, msg):
-        return {"role": "system", "content": msg}
+    def fsystem(self, msg: str) -> SystemMessage:
+        return SystemMessage(content=msg)
 
-    def fuser(self, msg):
-        return {"role": "user", "content": msg}
+    def fuser(self, msg: str) -> HumanMessage:
+        return HumanMessage(content=msg)
 
-    def fassistant(self, msg):
-        return {"role": "assistant", "content": msg}
+    def fassistant(self, msg: str) -> AIMessage:
+        return AIMessage(content=msg)
 
-    def next(self, messages: List[Dict[str, str]], prompt=None, *, step_name=None):
+    def combine_messages(self, messages: List[Message]) -> str:
+        msg_dict = messages_to_dict(messages)
+        logging.debug(msg_dict)
+        prompt = "\n".join(
+            "%s: %s" % (md["type"], md["data"]["content"]) for md in msg_dict
+        )
+        logging.debug("Prompt: " + prompt)
+        return prompt
+
+    def next(
+        self,
+        messages: List[Message],
+        prompt: Optional[str] = None,
+        *,
+        step_name: str,
+    ) -> List[Message]:
         if prompt:
-            messages += [{"role": "user", "content": prompt}]
+            messages.append(self.fuser(prompt))
 
         logger.debug(f"Creating a new chat completion: {messages}")
-        response = openai.ChatCompletion.create(
-            messages=messages,
-            stream=True,
-            model=self.model,
-            temperature=self.temperature,
-        )
 
-        chat = []
-        for chunk in response:
-            delta = chunk["choices"][0]["delta"]  # type: ignore
-            msg = delta.get("content", "")
-            print(msg, end="")
-            chat.append(msg)
-        print()
-        messages += [{"role": "assistant", "content": "".join(chat)}]
+        msgs_as_prompt = self.combine_messages(messages)
+        response = self.llm(msgs_as_prompt, callbacks=[StreamingStdOutCallbackHandler()])
+        response = cleanup_reponse(response)
+
+        messages.append(self.fassistant(response))
+
         logger.debug(f"Chat completion finished: {messages}")
 
         self.update_token_usage_log(
-            messages=messages, answer="".join(chat), step_name=step_name
+            messages=messages, answer=response, step_name=step_name
         )
 
         return messages
 
-    def update_token_usage_log(self, messages, answer, step_name):
+    def last_message_content(self, messages: List[Message]) -> str:
+        m = messages[-1].content.strip()
+        logging.info(m)
+        return m
+
+    @staticmethod
+    def serialize_messages(messages: List[Message]) -> str:
+        return json.dumps(messages_to_dict(messages))
+
+    @staticmethod
+    def deserialize_messages(jsondictstr: str) -> List[Message]:
+        return AI.parse_langchain_basemessages(
+            messages_from_dict(json.loads(jsondictstr))
+        )
```

Comment:
Can we just do:

`return list(messages_from_dict(json.loads(jsondictstr))` ?

Reply:
And delete below function

Reply:
Unfortunately no. Doing `list(...)` gives a `List[BaseMessage]`, but we need a `List[Message]` (which is defined as `List[Union[AIMessage, HumanMessage, SystemMessage]]`).

The complexity comes from the fact that `List` is [invariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)), meaning `List[Some_Subclass]` is not accepted where `List[Some_Class]` is expected.

For context:
- `BaseMessage` is a LangChain class and the superclass of the LC classes `AIMessage`, `HumanMessage`, and `SystemMessage`
- `Message` is defined in this PR and is an alias for `Union[AIMessage, HumanMessage, SystemMessage]`

We have to choose between one of these options:
1. Don't do typing at all
2. Use same type as LC: Use `Sequence[BaseMessage]` in gpt-engineer. This works because `Sequence` unlike `List` is covariant. But Sequence is immutable. So everywhere we add to the messages, we would have to convert to `List`, do the operation, then re-convert to `Sequence`. Imo this makes the code complex and repetitive.
3. Parse at the point of interaction with LC: Parse between LC's `Sequence[BaseMessage]` and our `List[Message]` only at the LC interface. Then we can keep our code simple.

I think 3 is cleanest, so I chose that.

What are your thoughts?

Reply:
Right the typing. My thinking is that we use typing.cast() (and assert if necessary)?

Otherwise the code is very confusing to the reader and I'd even prefer doing 1!
(which is a legit option: just adding  # type: ignore